File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/jsonengine/ShapeConverter.java
package com.wwdablu.soumya.simplypdf.jsonengine;

import android.graphics.Path;
import android.text.TextUtils;

import com.google.gson.Gson;
import com.wwdablu.soumya.simplypdf.composers.Composer;
import com.wwdablu.soumya.simplypdf.composers.ShapeComposer;
import com.wwdablu.soumya.simplypdf.composers.models.ShapeProperties;

import org.json.JSONArray;
import org.json.JSONObject;

class ShapeConverter extends BaseConverter {

    @Override
    protected void generate(Composer composer, JSONObject compose) throws Exception {

        if(!(composer instanceof ShapeComposer)) {
            return;
        }

        ShapeComposer shapeComposer = (ShapeComposer) composer;

        String shapePropertiesString = getProperties(compose);
        ShapeProperties shapeProperties = TextUtils.isEmpty(shapePropertiesString) ? null :
                new Gson().fromJson(shapePropertiesString, ShapeProperties.class);

        String shapeToDraw = compose.getString(Node.COMPOSER_SHAPE_SHAPE).toLowerCase();
        switch (shapeToDraw) {

            case "circle":
                drawCircle(shapeComposer, compose, shapeProperties);
                break;

            case "box":
                drawBox(shapeComposer, compose, shapeProperties);
                break;

            case "freeform":
                drawFreeform(shapeComposer, compose, shapeProperties);
                break;
        }
    }

    private void drawCircle(ShapeComposer shapeComposer,
                            JSONObject compose,
                            ShapeProperties shapeProperties) throws Exception {

        int radius = compose.getInt(Node.COMPOSER_SHAPE_RADIUS);
        shapeComposer.drawCircle(radius, shapeProperties);
    }

    private void drawBox(ShapeComposer shapeComposer,
                         JSONObject compose,
                         ShapeProperties shapeProperties) throws Exception {

        int width = compose.getInt(Node.COMPOSER_SHAPE_WIDTH);
        int height = compose.getInt(Node.COMPOSER_SHAPE_HEIGHT);

        shapeComposer.drawBox(width, height, shapeProperties);
    }

    private void drawFreeform(ShapeComposer shapeComposer, JSONObject compose, ShapeProperties shapeProperties) throws Exception {

        JSONArray pointsArray = compose.getJSONArray(Node.COMPOSER_SHAPE_POINTS);
        int pointsArrayLength = pointsArray.length();

        Path path = new Path();
        for(int pointIndex = 0; pointIndex < pointsArrayLength; pointIndex++) {

            JSONObject lineObject = pointsArray.getJSONObject(pointIndex);
            JSONArray lineArray = lineObject.getJSONArray(Node.COMPOSER_SHAPE_LINE);

            if(pointIndex == 0) {
                path.moveTo(lineArray.getInt(0), lineArray.getInt(1));
            } else {
                path.lineTo(lineArray.getInt(0), lineArray.getInt(1));
            }
        }

        shapeComposer.freeform(path, shapeProperties);
    }
}

File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/jsonengine/ShapeConverter.kt
package com.wwdablu.soumya.simplypdf.jsonengine

import android.graphics.Path
import android.text.TextUtils
import com.google.gson.Gson
import com.wwdablu.soumya.simplypdf.SimplyPdfDocument
import com.wwdablu.soumya.simplypdf.composers.ShapeComposer
import com.wwdablu.soumya.simplypdf.composers.properties.ShapeProperties
import com.wwdablu.soumya.simplypdf.jsonengine.base.ComposerConverter
import org.json.JSONObject

internal class ShapeConverter(simplyPdfDocument: SimplyPdfDocument) : ComposerConverter(simplyPdfDocument) {

    override fun generate(composeJsonObject: JSONObject) {

        val shapePropertiesString = getProperties(composeJsonObject)
        val shapeProperties = if (TextUtils.isEmpty(shapePropertiesString)) ShapeProperties()
            else Gson().fromJson(shapePropertiesString, ShapeProperties::class.java)

        val composer = simplyPdfDocument.shape

        when (composeJsonObject.getString(Node.COMPOSER_SHAPE_SHAPE).lowercase()) {
            "circle" -> drawCircle(composer, composeJsonObject, shapeProperties)
            "box" -> drawBox(composer, composeJsonObject, shapeProperties)
            "freeform" -> drawFreeform(composer, composeJsonObject, shapeProperties)
        }
    }

    private fun drawCircle(
        shapeComposer: ShapeComposer,
        compose: JSONObject,
        shapeProperties: ShapeProperties
    ) {

        val radius = compose.getInt(Node.COMPOSER_SHAPE_RADIUS).toFloat()
        shapeComposer.drawCircle(radius, shapeProperties)
    }

    private fun drawBox(
        shapeComposer: ShapeComposer,
        compose: JSONObject,
        shapeProperties: ShapeProperties
    ) {
        val width = compose.getInt(Node.COMPOSER_SHAPE_WIDTH)
        val height = compose.getInt(Node.COMPOSER_SHAPE_HEIGHT)
        shapeComposer.drawBox(width.toFloat(), height.toFloat(), shapeProperties)
    }

    private fun drawFreeform(
        shapeComposer: ShapeComposer,
        compose: JSONObject,
        shapeProperties: ShapeProperties
    ) {
        val pointsArray = compose.getJSONArray(Node.COMPOSER_SHAPE_POINTS)
        val pointsArrayLength = pointsArray.length()
        val path = Path()
        for (pointIndex in 0 until pointsArrayLength) {
            val lineObject = pointsArray.getJSONObject(pointIndex)
            val lineArray = lineObject.getJSONArray(Node.COMPOSER_SHAPE_LINE)
            if (pointIndex == 0) {
                path.moveTo(lineArray.getInt(0).toFloat(), lineArray.getInt(1).toFloat())
            } else {
                path.lineTo(lineArray.getInt(0).toFloat(), lineArray.getInt(1).toFloat())
            }
        }
        shapeComposer.freeform(path, shapeProperties)
    }

    override fun getTypeHandler(): String = "shape"
}